src/mame/nintendo/playch10.cpp

```diff
@@ -310,38 +629,850 @@ WRITE_LINE_MEMBER(playch10_state::up8w_w)
+
+	// default mirroring
+	m_mirroring = PPU_MIRROR_VERT;
+}
+
+/**********************************************************************************
+ *
+ *  Game and Board-specific initialization
+ *
+ **********************************************************************************/
+
+// Gun games
+
+void playch10_state::init_pc_gun()
+{
+	// common init
+	init_playch10();
+
+	// set the control type
+	m_pc10_gun_controller = 1;
+}
+
+// Horizontal mirroring
+
+void playch10_state::init_pc_hrz()
+{
+	// common init
+	init_playch10();
+
+	// setup mirroring
+	m_mirroring = PPU_MIRROR_HORZ;
+}
+
+// Init for games that bank PRG memory
+
+void playch10_state::init_prg_banking()
+{
+	// common init
+	init_playch10();
+
+	u8 *base = memregion("prg")->base();
+	m_prg_chunks = memregion("prg")->bytes() / 0x2000;
+
+	for (int i = 0; i < 4; i++)
+	{
+		m_prg_banks[i]->configure_entries(0, m_prg_chunks, base, 0x2000);
+		m_prg_banks[i]->set_entry(m_prg_chunks - 4 + i);
+	}
+
+	m_prg_view.select(0);
+}
+
+// safe banking helpers (only work when PRG size is a power of 2)
+
+void playch10_state::prg32(int bank)
+{
+	bank = (bank << 2) & (m_prg_chunks - 1);
+
+	for (int i = 0; i < 4; i++)
+		m_prg_banks[i]->set_entry(bank + i);
+}
+
+void playch10_state::prg16(int slot, int bank)
+{
+	bank = (bank << 1) & (m_prg_chunks - 1);
+	slot = (slot & 1) << 1;
+
+	for (int i = 0; i < 2; i++)
+		m_prg_banks[slot + i]->set_entry(bank + i);
+}
+
+void playch10_state::prg8(int slot, int bank)
+{
+	m_prg_banks[slot & 0x03]->set_entry(bank & (m_prg_chunks - 1));
+}
+
+// MMC1 mapper, used by D, F, and K boards
+
+void playch10_state::mmc1_rom_switch_w(offs_t offset, u8 data)
+{
+	static int size16k, switchlow, vrom4k;
+
+	// reset mapper
+	if (data & 0x80)
+	{
+		m_mmc1_shiftreg = m_mmc1_shiftcount = 0;
+
+		size16k = 1;
+		switchlow = 1;
+		vrom4k = 0;
+
+		return;
+	}
+
+	// update shift register
+	m_mmc1_shiftreg = (m_mmc1_shiftreg >> 1) | (data & 1) << 4;
+	m_mmc1_shiftcount = (m_mmc1_shiftcount + 1) % 5;
+
+	// are we done shifting?
+	if (!m_mmc1_shiftcount)
+	{
+		// apply data to registers
+		switch (BIT(offset, 13, 2))
+		{
+			case 0: // mirroring and options
+			{
+				static constexpr u8 mirr[4] = {
+					PPU_MIRROR_LOW, PPU_MIRROR_HIGH, PPU_MIRROR_VERT, PPU_MIRROR_HORZ
+				};
+
+				pc10_set_mirroring(mirr[m_mmc1_shiftreg & 0x03]);
+
+				vrom4k = m_mmc1_shiftreg & 0x10;
+				size16k = m_mmc1_shiftreg & 0x08;
+				switchlow = m_mmc1_shiftreg & 0x04;
+				break;
+			}
+
+			case 1: // video rom banking - bank 0 - 4k or 8k
+				if (m_vram)
+					set_videoram_bank(0, (vrom4k) ? 4 : 8, m_mmc1_shiftreg & 0x1f, 4);
+				else
+					pc10_set_videorom_bank(0, (vrom4k) ? 4 : 8, m_mmc1_shiftreg & 0x1f, 4);
+				break;
+
+			case 2: // video rom banking - bank 1 - 4k only
+				if (vrom4k)
+				{
+					if (m_vram)
+						set_videoram_bank(0, (vrom4k) ? 4 : 8, m_mmc1_shiftreg & 0x1f, 4);
+					else
+						pc10_set_videorom_bank(4, 4, m_mmc1_shiftreg & 0x1f, 4);
+				}
```

Comment:
Coverity has helpfully pointed out that the inner `(vrok4k)` test on line 1108 can never be false because the outer `vrom4k` test on line 1105 would prevent it from getting there.

Reply:
Thanks. I've been avoiding cleaning the multiple implementations of MMC1/2/3 up too much, thinking these PC10 etc ones will be merged with something better at some point. But since I'm aware there's another bug just a few lines above (I think) maybe I'll put this on the near-term todo list.